File: 2012/July/projects/skeleton/Lexicon/ex48/lexicon.py
##lexicon script for exercise 48

# scanner to analyse lexicon
lexicon = {'north':'direction', 'south':'direction', 'east':'direction',
'west':'direction', 'down':'direction', 'up':'direction', 'left':'direction'
,'right':'direction', 'back':'direction', 'go':'verb', 'stop':'verb',
'kill':'verb', 'eat':'verb', 'the':'stop', 'in':'stop', 'of':'stop',
'from':'stop', 'at':'stop', 'it':'stop', 'door':'noun', 'bear':'noun',
'princess':'noun', 'cabinet':'noun'}

def scan(instr):
	inlist = instr.lower().split()
	outlist = []
	for item in inlist:
		if item.isdigit():
			outlist.append(('number', int(item)))
		else:
			outlist.append((lexicon.get(item,'error'), item))
	return outlist

File: Learn_Ruby_The_Hard_Way/ex48/lib/lexicon.rb
Pair = Struct.new(:token, :word)

class Lexicon
	def initialize()
		@direction = ["north", "south", "east", "west", "down", "up",
			"left", "right", "back"]
		@verb = ["go", "stop", "kill", "eat"]
		@stop = ["the", "in", "of", "from", "at", "it"]
		@noun = ["door", "bear", "princess", "cabinet"]
	end

	def scan(arg)
		words = arg.split(" ")
		sentence = []
		words.each do |i|
			if @direction.include?(i)
				i = Pair.new(:direction, i)
				sentence.push(i)
			elsif @verb.include?(i)
				i = Pair.new(:verb, i)
				sentence.push(i)
			elsif @stop.include?(i)
				i = Pair.new(:stop, i)
				sentence.push(i)
			elsif @noun.include?(i)
				i = Pair.new(:noun, i)
				sentence.push(i)
			else
				begin
					i = Integer(i)
					sentence.push(Pair.new(:number, i))
				rescue
					i = Pair.new(:error, i)
					sentence.push(i)
				end
			end
		end
		return sentence
	end
end
